### DISCUSSION TITLE
Liquid Access to Query Params

### DISCUSSION THREAD

**Discussion by dhruveonmars** [Shopify/liquid#1872] [Requests for improvements]
Would it be possible to get access to URL params in Liquid?

Currently, a lot of themes, either need to use JS to adjust the page/DOM after it's loaded based on query params. Some others use `content_for_header` hacks, however neither are great solutions. `content_for_header` can lead to bugs, can change at any point and can be unpredictable. JS means loading in all content, all the time, even if not needed or requires additional fetch requests to load appropriate items. This creates an issue with performance.

Would be great if in the `request` object, we could add a property like `query_string`. This is currently available to Shopify Pixels, but not the theme.

**Comment by bakura10** [comment]
I agree this would be awesome, and would remove some hacks we’re doing (such as storing thing in cart attributes). The issue is that it could lead to a lot of cache miss as potentially any query param could invalidate the cache and could be a vector attack for shopify (generate millions of requests with different query parameters to bypass the cache layer).

maybe we could have a new parameter in the settings that would allow a dev to whitelist a specific list of query params.

**Reply by dhruveonmars** [reply]
Good point @bakura10! 

So maybe an object, like `query_params` in `settings_data`, or a new file? I think it would just need to be an array of allowed params.
Eg
```
["faq_search", "some_other_key"]
```

would then allow us to retrieve this value in Liquid
eg URL store.com/pages/help?faq_search=shipping
`shop.query['faq_search']` or `shop.query.faq_search` would return `shipping`

**Comment by dhruveonmars** [comment]
@benjaminsehl Any thoughts on whether this is something that could be added or looked into?

**Comment by benjaminsehl** [comment]
@dhruveonmars @bakura10 — sorry I missed this way back, but this convo reared it's head again — I'd be curious to know what some example use cases are that you reach for most often?

**Reply by dhruveonmars** [reply]
@benjaminsehl We currently have 3 main use cases.

1. Searching metaobjects, for eg on a help center / faq page. Each category is a metaobject entry of type 'help center category', which has a field for a list of questions/answers, which is another metaobject of type 'help center item'. 
  - We previously used content for header hack to do this. When that had issues, we changed to using a hack using the native search page, section rendering API and a liquid template. Another previous solution for this, was to load everything in the page and use JS to filter the data. Again, far from ideal, and performance ends up being a trade-off. Also means when urls are shared directly, eg store.com/pages/help?q=where+is+my+order, we can't use Liquid to load in things correctly on page load.

Ideally we would be able to use section rendering API and the theme can use these natively in Liquid. 

2. We have some custom products, which result in over 6k variants on average, which currently isn't possible natively. Customers can choose a frame of a video they wish to buy. Because we don't have this value in liquid, which in the url will normally be something like `?frame=123`, we have scripts to read this value and then change the PDP. We have to change the images after page load, change the information, eg on the page, it shows 'Frame 123', where this value changes. Similarly, because each frame number has a different image, we have to set those images, after the url search params are retrieved.

3. We also have dynamic redirects on login - eg for subscriptions, auctions, and a few others features, which redirect people back to the previous page, or specific products upon login. This uses url params as well, which then need JS to change the value of form `return_to`

**Comment by Muhammad-Hasham-Khalid** [comment]
This would be awesome, we also have a use-case of metaobjects searching through query params, if this finds its way to the tooling it would be amazing.

**Comment by MatthewRCrigger** [comment]
I think for caching reasons maybe Shopify may not want to pay attention to and send query params. However `content_for_header` can leak them, with a slim chance of getting an outdated value. In addition to content for header hacking, I noticed the `routes` `storefront_login_url` has a param of `return_to` which contains the url encoded landed URL with params and all. So I have just been breaking that apart and grabbing query params that way for the time being.

Of course no one wants to hack this stuff, so for now I'm making up reasons why this probably isn't available.
